src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -174,10 +174,6 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
 
             private void OnDiagnosticsUpdated(DiagnosticsUpdatedArgs e)
             {
-                // First see if this is a document/project removal.  If so, clear out any state we
-                // have associated with any analyzers we have for that document/project.
-                ProcessRemovedDiagnostics(e);
```

Comment:
I see that 1. is not possible because the editor asks us to create a tagger per buffer [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs,55). Additionally, the [TagsChanged](http://index/#Microsoft.VisualStudio.Text.Logic/Tagging/ITagger.cs,d12b0ec733ef2a2e,references) event is also defined in the editor layer, so we can't include workspace information. I think best we can do is take approach 2 above, and invoke TagsChanged only when we receive a diagnostics updated event for subject buffer's current owning workspace.